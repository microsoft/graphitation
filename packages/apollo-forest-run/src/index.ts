export { ForestRun } from "./ForestRun";
export { ForestRunCompat } from "./ForestRunCompat";
