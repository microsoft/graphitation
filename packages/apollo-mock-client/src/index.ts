import {
  ApolloLink,
  Observable,
  ApolloClient,
  InMemoryCache,
} from "@apollo/client";
import type {
  ApolloCache,
  Operation,
  FetchResult,
  NormalizedCacheObject,
  InMemoryCacheConfig,
} from "@apollo/client";
import type { SubscriptionObserver } from "zen-observable-ts";
import { assertType, isAbstractType } from "graphql";
import type { DocumentNode, ExecutionResult, GraphQLSchema } from "graphql";
import invariant from "invariant";

export interface RequestDescriptor<Node = DocumentNode> {
  readonly node: Node;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  readonly variables: Record<string, any>;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  readonly context?: Record<string, any>;
}

export interface OperationDescriptor<
  Schema = GraphQLSchema,
  Node = DocumentNode,
> {
  readonly schema: Schema;
  readonly request: RequestDescriptor<Node>;
}

type OperationMockResolver<Schema = GraphQLSchema, Node = DocumentNode> = (
  operation: OperationDescriptor<Schema, Node>,
) => ExecutionResult | Error | undefined | null;

export interface MockFunctions<Schema = GraphQLSchema, Node = DocumentNode> {
  /**
   * Get all operation executed during the test by the current time.
   */
  getAllOperations(): OperationDescriptor<Schema, Node>[];

  /**
   * Return the most recent operation. This method will throw if no operations were executed prior this call.
   */
  getMostRecentOperation(): OperationDescriptor<Schema, Node>;

  /**
   * Find a particular operation in the list of all executed operations. This method will throw if the operation is not
   * found.
   */
  findOperation(
    findFn: (operation: OperationDescriptor<Schema, Node>) => boolean,
  ): OperationDescriptor<Schema, Node>;

  /**
   * Provide a payload for an operation, but not complete the request. Practically useful when testing incremental
   * updates and subscriptions.
   *
   * @note
   *
   * ApolloClient requires a delay until the next tick of the runloop before it updates,
   * as per https://www.apollographql.com/docs/react/development-testing/testing/
   */
  nextValue(
    operation: OperationDescriptor<Schema, Node>,
    data: ExecutionResult,
  ): Promise<void>;

  /**
   * Complete the operation. No more payloads are expected for this operation.
   *
   * @note
   *
   * ApolloClient requires a delay until the next tick of the runloop before it updates,
   * as per https://www.apollographql.com/docs/react/development-testing/testing/
   */
  complete(operation: OperationDescriptor<Schema, Node>): Promise<void>;

  /**
   * Resolve the request with the provided payload. This is a shortcut for `nextValue(...)` and `complete(...)`.
   *
   * @note
   *
   * ApolloClient requires a delay until the next tick of the runloop before it updates,
   * as per https://www.apollographql.com/docs/react/development-testing/testing/
   */
  resolve(
    operation: OperationDescriptor<Schema, Node>,
    data: ExecutionResult,
  ): Promise<void>;

  /**
   * Reject the request with a given error.
   *
   * @note
   *
   * ApolloClient requires a delay until the next tick of the runloop before it updates,
   * as per https://www.apollographql.com/docs/react/development-testing/testing/
   */
  reject(
    operation: OperationDescriptor<Schema, Node>,
    error: Error,
  ): Promise<void>;

  /**
   * A shortcut for `getMostRecentOperation()` and `resolve()`.
   *
   * @note
   *
   * ApolloClient requires a delay until the next tick of the runloop before it updates,
   * as per https://www.apollographql.com/docs/react/development-testing/testing/
   */
  resolveMostRecentOperation(
    resolver: (operation: OperationDescriptor<Schema, Node>) => ExecutionResult,
  ): Promise<void>;

  /**
   * A shortcut for `getMostRecentOperation()` and `reject()`.
   * @note
   *
   * ApolloClient requires a delay until the next tick of the runloop before it updates,
   * as per https://www.apollographql.com/docs/react/development-testing/testing/
   */
  rejectMostRecentOperation(
    error: Error | ((operation: OperationDescriptor<Schema, Node>) => Error),
  ): Promise<void>;

  /**
   * Adds a resolver function that will be used to resolve/reject operations as they appear.
   */
  queueOperationResolver: (
    resolver: OperationMockResolver<Schema, Node>,
  ) => Promise<void>;
}

interface ApolloClientExtension {
  mock: MockFunctions;
  // mockClear: () => void;
}

export interface ApolloMockClient
  extends ApolloClient<NormalizedCacheObject>,
    ApolloClientExtension {}

class MockLink extends ApolloLink {
  public schema: GraphQLSchema;
  public mock: Mock;

  constructor(schema: GraphQLSchema) {
    super();
    this.schema = schema;
    this.mock = new Mock();
  }

  // FIXME: This does't actually work well and is likely due to the client
  //        being in a tainted state.
  // public mockClear() {
  //   this.mock = new Mock();
  // }

  public request(operation: Operation): Observable<FetchResult> | null {
    return new Observable<FetchResult>((observer) => {
      this.mock.addOperation(
        {
          schema: this.schema,
          request: {
            node: operation.query,
            variables: operation.variables || {},
            context: operation.getContext(),
          },
        },
        observer,
      );
    });
  }
}

function executeOperationMockResolver(
  resolver: OperationMockResolver,
  operation: OperationDescriptor,
  observer: SubscriptionObserver<FetchResult>,
) {
  const resolved = resolver(operation);
  if (resolved) {
    if (resolved instanceof Error) {
      observer.error(resolved);
    } else {
      observer.next(resolved);
    }
    observer.complete();
    return true;
  }
  return false;
}

class Mock implements MockFunctions {
  private operations: Map<
    OperationDescriptor,
    SubscriptionObserver<FetchResult>
  >;

  private resolversQueue: OperationMockResolver[];

  constructor() {
    this.operations = new Map();
    this.resolversQueue = [];
  }

  public addOperation(
    operation: OperationDescriptor,
    observer: SubscriptionObserver<FetchResult>,
  ) {
    for (const resolver of this.resolversQueue) {
      if (executeOperationMockResolver(resolver, operation, observer)) {
        return;
      }
    }
    // If not immediately resolved, store it for later
    this.operations.set(operation, observer);
  }

  private getObserver(operation: OperationDescriptor) {
    const observer = this.operations.get(operation);
    invariant(observer, "Could not find operation in execution queue");
    return observer;
  }

  /**
   * MockFunctions
   */

  public getAllOperations(): OperationDescriptor[] {
    return Array.from(this.operations.keys());
  }

  public getMostRecentOperation(): OperationDescriptor {
    const operations = this.getAllOperations();
    invariant(
      operations.length > 0,
      "Expected at least one operation to have been started",
    );
    return operations[operations.length - 1];
  }

  public findOperation(
    findFn: (operation: OperationDescriptor) => boolean,
  ): OperationDescriptor {
    let result: OperationDescriptor | null = null;
    for (const operation of this.operations.keys()) {
      if (findFn(operation)) {
        result = operation;
        break;
      }
    }
    invariant(
      result,
      "Operation was not found in the list of pending operations",
    );
    return result;
  }

  public async nextValue(
    operation: OperationDescriptor,
    data: ExecutionResult,
  ): Promise<void> {
    this.getObserver(operation).next(data);
  }

  public async complete(operation: OperationDescriptor): Promise<void> {
    const observer = this.getObserver(operation);
    observer.complete();
    this.operations.delete(operation);
  }

  public async resolve(
    operation: OperationDescriptor,
    data: ExecutionResult,
  ): Promise<void> {
    this.nextValue(operation, data);
    this.complete(operation);
  }

  public async reject(
    operation: OperationDescriptor,
    error: Error,
  ): Promise<void> {
    this.getObserver(operation).error(error);
    this.complete(operation);
  }

  public async resolveMostRecentOperation(
    resolver: (operation: OperationDescriptor) => ExecutionResult,
  ): Promise<void> {
    const operation = this.getMostRecentOperation();
    this.resolve(operation, resolver(operation));
  }

  public async rejectMostRecentOperation(
    error: Error | ((operation: OperationDescriptor) => Error),
  ): Promise<void> {
    const operation = this.getMostRecentOperation();
    this.reject(
      operation,
      typeof error === "function" ? error(operation) : error,
    );
  }

  public async queueOperationResolver(resolver: OperationMockResolver) {
    this.resolversQueue.push(resolver);
    for (const [operation, observer] of this.operations) {
      if (executeOperationMockResolver(resolver, operation, observer)) {
        this.operations.delete(operation);
      }
    }
  }
}

export function createMockClient(
  schema: GraphQLSchema,
  options?: {
    cache?: ApolloCache<any>;
    inMemoryCacheConfig?: InMemoryCacheConfig;
  },
): ApolloMockClient {
  let cache = options?.cache;
  if (!cache) {
    // Build a list of abstract types and their possible types.
    // TODO: Cache this on the schema?
    const possibleTypes: Record<string, string[]> = {};
    Object.keys(schema.getTypeMap()).forEach((typeName) => {
      const type = schema.getType(typeName);
      assertType(type);
      if (isAbstractType(type)) {
        possibleTypes[typeName] = schema
          .getPossibleTypes(type)
          .map((possibleType) => possibleType.name);
      }
    });
    cache = new InMemoryCache({
      addTypename: true,
      possibleTypes,
      ...options?.inMemoryCacheConfig,
    });
  }

  const link = new MockLink(schema);

  return Object.assign<
    ApolloClient<NormalizedCacheObject>,
    ApolloClientExtension
  >(
    new ApolloClient({
      cache,
      link,
    }),
    {
      get mock() {
        return link.mock;
      },
      // mockClear() {
      //   link.mockClear();
      // },
    },
  );
}
