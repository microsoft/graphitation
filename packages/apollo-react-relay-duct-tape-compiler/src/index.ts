export * from "./typescriptTransforms/createImportDocumentsTransform";
export * from "./types";
