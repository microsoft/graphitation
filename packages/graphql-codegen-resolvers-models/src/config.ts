import { RawConfig } from "@graphql-codegen/visitor-plugin-common";

export type MapperConfigValue = {
  extend?: boolean;
  exclude?: string[];
};
export interface ResolversModelsPluginConfig extends RawConfig {
  mappers?: { [typeName: string]: string };
  mapperTypeSuffix?: string;
  mappersConfig?: {
    [typeName: string]: MapperConfigValue;
  };
  federation?: boolean;
  modelIntersectionSuffix?: string;
  namespacedImportName?: string;
}
