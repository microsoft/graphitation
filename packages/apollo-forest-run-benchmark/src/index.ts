import type { CacheConfig, WorkerResult, SuiteRawResult } from "./types";

import path from "path";
import { CACHE_FACTORIES } from "./config";
import { log } from "./utils/logger";
import { analyzeResults } from "./summary/analyze-results";
import { CONFIG } from "./config";
import { spawn } from "child_process";
import { mergeResults } from "./utils/merge";
import { getSummary } from "./summary/summary";

interface BaseSuite {
  cacheFactory: (typeof CACHE_FACTORIES)[number];
  cacheConfig: CacheConfig;
}

const BASE_SUITES: BaseSuite[] = [];

for (const cacheFactory of CACHE_FACTORIES) {
  for (const cacheConfig of CONFIG.cacheConfigurations) {
    BASE_SUITES.push({ cacheFactory, cacheConfig });
  }
}

const spawnProcess = (job: BaseSuite): Promise<WorkerResult> => {
  return new Promise((resolve) => {
    const workerScript = path.join(__dirname, "runners", "suite-worker.js");
    const child = spawn(
      process.execPath,
      [
        "--noconcurrent_sweeping",
        "--noconcurrent_recompilation",
        workerScript,
        JSON.stringify(job),
      ],
      {
        env: {
          ...process.env,
        },
        stdio: ["pipe", "pipe", "pipe"],
      },
    );

    let stdout = "";

    child.stdout.on("data", (data) => {
      stdout += data.toString();
    });

    child.stderr.on("data", (data) => {
      console.error(data.toString());
    });

    child.on("close", () => {
      const results = JSON.parse(stdout.trim());
      resolve(results);
    });
  });
};

const runBaseSuites = async (): Promise<WorkerResult[]> => {
  const allResults: WorkerResult[] = [];

  for (const suite of BASE_SUITES) {
    console.log(
      `\n=== Running benchmarks for ${suite.cacheFactory.name} with ${suite.cacheConfig.name} in isolated process ===`,
    );

    const result = await spawnProcess(suite);
    allResults.push(result);
    console.log(
      `Completed ${suite.cacheFactory.name}/${suite.cacheConfig.name} with ${result.results.length} results`,
    );
  }

  return allResults;
};

const runBenchmarks = async (): Promise<void> => {
  const { epochs } = CONFIG.reliability;
  const prevSuites: SuiteRawResult[] = [];
  log.start();
  for (let attempt = 1; attempt <= epochs; attempt++) {
    log.attempt(attempt);

    const results = await runBaseSuites();
    const groupedResult = mergeResults(results);

    prevSuites.push(groupedResult);
  }

  const summary = getSummary(prevSuites);
  analyzeResults(summary);
};

runBenchmarks();
