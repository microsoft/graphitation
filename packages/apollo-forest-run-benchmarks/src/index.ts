import { CACHE_FACTORIES, CacheConfig } from "./config";
import fs from "fs";
import { isResultReliable, groupResults, getSummary } from "./reliability";
import { log } from "./utils/logger";
import { analyzeResults } from "./analyze-results";
import { CONFIG } from "./config";
import { scenarios } from "./scenarios";
import { spawn } from "child_process";
import path from "path";
import { garbageCollect } from "./utils/garbage-collection";

export interface ResultIdentifier {
  cacheConfig: CacheConfig["name"];
  cacheFactory: string;
}

export interface Result extends ResultIdentifier {
  scenario: `${(typeof scenarios)[number]["name"]}_${number}`;
  samples: number[];
  operationName: string;
}

interface BenchmarkJob {
  cacheFactory: (typeof CACHE_FACTORIES)[number];
  cacheConfig: CacheConfig;
}

const benchmarkJobs: BenchmarkJob[] = [];
for (const cacheFactory of CACHE_FACTORIES) {
  for (const cacheConfig of CONFIG.cacheConfigurations) {
    benchmarkJobs.push({ cacheFactory, cacheConfig });
  }
}

function runBenchmarkInIsolatedProcess(job: BenchmarkJob): Promise<Result[]> {
  return new Promise((resolve) => {
    const workerScript = path.join(__dirname, "benchmark-worker.ts");
    const child = spawn(
      process.execPath,
      ["-r", "ts-node/register", workerScript, JSON.stringify(job)],
      {
        env: {
          ...process.env,
          TS_NODE_COMPILER_OPTIONS: '{"module":"commonjs"}',
        },
        stdio: ["pipe", "pipe", "pipe"],
      },
    );

    let stdout = "";

    child.stdout.on("data", (data) => {
      stdout += data.toString();
    });

    child.stderr.on("data", (data) => {
      console.error(`Worker stderr: ${data}`);
    });

    child.on("close", () => {
      const results = JSON.parse(stdout.trim());
      resolve(results);
    });
  });
}

async function runBenchmarkSuite(): Promise<Result[]> {
  const allResults: Result[] = [];

  for (const job of benchmarkJobs) {
    console.log(
      `\n=== Running benchmarks for ${job.cacheFactory.name} with ${job.cacheConfig.name} in isolated process ===`,
    );

    const jobResults = await runBenchmarkInIsolatedProcess(job);
    allResults.push(...jobResults);
    console.log(
      `Completed ${job.cacheFactory.name}/${job.cacheConfig.name} with ${jobResults.length} results`,
    );
  }

  return allResults;
}

export interface BenchmarkResult {
  [scenarioName: string]: Result[];
}

async function runBenchmarks(): Promise<void> {
  const { maxAttempts } = CONFIG.reliability;
  let prevBenchmarks: BenchmarkResult[] = [];
  log.start();
  for (let attempt = 1; attempt <= maxAttempts; attempt++) {
    log.attempt(attempt);

    const currentResult = await runBenchmarkSuite();
    const groupedResults = groupResults(currentResult);
    const isReliable = isResultReliable(groupedResults, prevBenchmarks);

    prevBenchmarks.push(groupedResults);

    if (isReliable && attempt > CONFIG.reliability.minAttempts) {
      break;
    }
  }

  const summary = getSummary(prevBenchmarks);
  analyzeResults(summary);
  fs.writeFileSync("benchmark-summary.json", JSON.stringify(summary, null, 2));
}

runBenchmarks();
