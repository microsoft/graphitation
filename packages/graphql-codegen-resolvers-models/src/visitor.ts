import { ResolversModelsPluginConfig } from "./config";
import { GraphQLSchema } from "graphql";
import { ApolloFederation } from "@graphql-codegen/plugin-helpers";

import autoBind from "auto-bind";
import {
  ParsedResolversConfig,
  BaseVisitor,
  getConfigValue,
  transformMappers,
  ParsedMapper,
} from "@graphql-codegen/visitor-plugin-common";

export class ResolversModelsVisitor<
  TRawConfig extends ResolversModelsPluginConfig = ResolversModelsPluginConfig,
  TPluginConfig extends ParsedResolversConfig = ParsedResolversConfig,
> extends BaseVisitor<TRawConfig, TPluginConfig> {
  protected _federation: ApolloFederation;

  constructor(rawConfig: TRawConfig, private schema: GraphQLSchema) {
    super(rawConfig, {
      federation: getConfigValue(rawConfig.federation, false),
      mappers: transformMappers(
        rawConfig.mappers || {},
        rawConfig.mapperTypeSuffix,
      ),
    } as TPluginConfig);

    autoBind(this);
    this._federation = new ApolloFederation({
      enabled: this.config.federation,
      schema: this.schema,
    });
  }

  public getValidMappers() {
    const allSchemaTypes = this.schema.getTypeMap();
    const typeNames = this._federation.filterTypeNames(
      Object.keys(allSchemaTypes),
    );
    return typeNames.reduce((acc, typeName) => {
      const isMapped = this.config.mappers[typeName];
      if (isMapped) {
        acc[typeName] = this.config.mappers[typeName];
      }
      return acc;
    }, {} as { [typename: string]: ParsedMapper });
  }
}
