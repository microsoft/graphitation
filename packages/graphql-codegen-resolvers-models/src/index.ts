import { printSchemaWithDirectives } from "@graphql-tools/utils";
import ts, { factory, TypeAliasDeclaration } from "typescript";
import { ParsedMapper } from "@graphql-codegen/visitor-plugin-common";
import {
  Types,
  PluginFunction,
  addFederationReferencesToSchema,
} from "@graphql-codegen/plugin-helpers";
import { parse, visit, GraphQLSchema, printSchema } from "graphql";
import { ResolversModelsVisitor } from "./visitor";
import { ResolversModelsPluginConfig, MapperConfigValue } from "./config";

export const plugin: PluginFunction<
  ResolversModelsPluginConfig,
  Types.ComplexPluginOutput
> = (
  schema: GraphQLSchema,
  documents: Types.DocumentFile[],
  config: ResolversModelsPluginConfig,
) => {
  const transformedSchema = config.federation
    ? addFederationReferencesToSchema(schema)
    : schema;

  if (!config.modelIntersectionSuffix || !config.mappers) {
    return { content: "" };
  }

  const visitor = new ResolversModelsVisitor(config, transformedSchema);

  const printedSchema = config.federation
    ? printSchemaWithDirectives(transformedSchema)
    : printSchema(transformedSchema);
  const astNode = parse(printedSchema);
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  visit(astNode, { leave: visitor } as any);

  const content = Object.entries(visitor.getValidMappers()).reduce(
    (acc, [typeName, model]) => {
      acc.push(
        ...getModelAST(
          typeName,
          model,
          config.modelIntersectionSuffix as string,
          config.namespacedImportName || "",
          config?.mappersConfig?.[typeName],
        ),
      );
      return acc;
    },
    [] as TypeAliasDeclaration[],
  );

  const tsContents = factory.createSourceFile(
    content,
    factory.createToken(ts.SyntaxKind.EndOfFileToken),
    0,
  );

  const printer = ts.createPrinter();

  return {
    content: printer.printNode(ts.EmitHint.SourceFile, tsContents, tsContents),
  };
};

function getOmittedFields(typeName: string, omitFields?: string[]) {
  if (!omitFields || !omitFields.length) {
    return;
  }
  return factory.createTypeAliasDeclaration(
    undefined,
    factory.createIdentifier(`${typeName}ModelOmitFields`),
    undefined,
    factory.createUnionTypeNode(
      omitFields.map((field) =>
        factory.createLiteralTypeNode(factory.createStringLiteral(field)),
      ),
    ),
  );
}

function getModelAST(
  typeName: string,
  mapper: ParsedMapper,
  modelIntersectionSuffix: string,
  namespacedImportName: string,
  mapperConfig?: MapperConfigValue,
) {
  if (!mapperConfig || !mapperConfig.extend) {
    return [
      factory.createTypeAliasDeclaration(
        [factory.createModifier(ts.SyntaxKind.ExportKeyword)],
        factory.createIdentifier(mapper.type),
        undefined,
        factory.createTypeReferenceNode(
          factory.createIdentifier(`${mapper.type}${modelIntersectionSuffix}`),
          undefined,
        ),
      ),
    ];
  }

  const omittedFields = getOmittedFields(typeName, mapperConfig.exclude);

  if (!omittedFields) {
    return [
      factory.createTypeAliasDeclaration(
        [factory.createModifier(ts.SyntaxKind.ExportKeyword)],
        factory.createIdentifier(mapper.type),
        undefined,
        factory.createIntersectionTypeNode([
          factory.createTypeReferenceNode(
            factory.createIdentifier(
              `${
                namespacedImportName ? namespacedImportName + "." : ""
              }${typeName}`,
            ),
            undefined,
          ),
          factory.createTypeReferenceNode(
            factory.createIdentifier(
              `${mapper.type}${modelIntersectionSuffix}`,
            ),
            undefined,
          ),
        ]),
      ),
    ];
  }

  return [
    omittedFields,
    factory.createTypeAliasDeclaration(
      [factory.createModifier(ts.SyntaxKind.ExportKeyword)],
      factory.createIdentifier(mapper.type),
      undefined,
      factory.createIntersectionTypeNode([
        factory.createTypeReferenceNode(factory.createIdentifier("Omit"), [
          factory.createTypeReferenceNode(
            factory.createIdentifier(
              `${
                namespacedImportName ? namespacedImportName + "." : ""
              }${typeName}`,
            ),
            undefined,
          ),
          factory.createTypeReferenceNode(
            factory.createIdentifier(`${typeName}ModelOmitFields`),
            undefined,
          ),
        ]),
        factory.createTypeReferenceNode(
          factory.createIdentifier(`${mapper.type}${modelIntersectionSuffix}`),
          undefined,
        ),
      ]),
    ),
  ];
}

export { ResolversModelsVisitor };
