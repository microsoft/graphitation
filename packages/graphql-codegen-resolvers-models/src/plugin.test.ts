import { buildSchema } from "graphql";
import { plugin } from "../src";

const testSchema = buildSchema(/* GraphQL */ `
  schema {
    query: Query
  }

  type Query {
    role: [ProjectRoleDetail!]!
  }

  type ProjectRoleDetail {
    code: String!
    name: String!
    email: String!
  }
`);

describe("TypeScript Resolvers Plugin", () => {
  it("generates models with omitted fields", async () => {
    const config = {
      modelIntersectionSuffix: "Template",
      mappersConfig: {
        ProjectRoleDetail: {
          extend: true,
          exclude: ["name", "email"],
        },
      },
      mappers: {
        ProjectRoleDetail: "../entities#ProjectRole",
      },
    };
    const output = await plugin(testSchema, [], config, {
      outputFile: "graphql.ts",
    });

    expect(output.content).toMatchSnapshot();
  });

  it("generates the models without omitted fields", async () => {
    const config = {
      modelIntersectionSuffix: "Template",
      mappersConfig: {
        ProjectRoleDetail: {
          extend: true,
        },
      },
      mappers: {
        ProjectRoleDetail: "../entities#ProjectRole",
      },
    };
    const output = await plugin(testSchema, [], config, {
      outputFile: "graphql.ts",
    });

    expect(output.content).toMatchSnapshot();
  });

  it("just assigns alias into the type, which is used in resolverTypes", async () => {
    const config = {
      modelIntersectionSuffix: "Template",
      mappers: {
        ProjectRoleDetail: "../entities#ProjectRole",
      },
    };
    const output = await plugin(testSchema, [], config, {
      outputFile: "graphql.ts",
    });

    expect(output.content).toMatchSnapshot();
  });

  it("doesn't generate anything", async () => {
    const baseConfig = {
      mappersConfig: {
        ProjectRoleDetail: {
          extend: true,
        },
      },
    };
    expect(
      ((await plugin(
        testSchema,
        [],
        {
          ...baseConfig,
          mappers: {
            ProjectRoleDetail: "../entities#ProjectRole",
          },
        },
        {
          outputFile: "graphql.ts",
        },
      )) as any).content,
    ).toEqual("");

    expect(
      ((await plugin(
        testSchema,
        [],
        {
          ...baseConfig,
          modelIntersectionSuffix: "Template",
        },
        {
          outputFile: "graphql.ts",
        },
      )) as any).content,
    ).toEqual("");
  });
});
