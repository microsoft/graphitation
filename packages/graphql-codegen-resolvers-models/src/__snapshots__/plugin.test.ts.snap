// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypeScript Resolvers Plugin Should generate the models with omit 1`] = `
"type ProjectRoleDetailModelOmitFields = \\"name\\" | \\"email\\";
type ProjectRole = Omit<ProjectRoleDetail, ProjectRoleDetailModelOmitFields> & ProjectRoleTemplate;
"
`;

exports[`TypeScript Resolvers Plugin Should generate the models without omit 1`] = `
"type ProjectRole = ProjectRoleDetail & ProjectRoleTemplate;
"
`;

exports[`TypeScript Resolvers Plugin Should just assign alias into the type 1`] = `
"type ProjectRole = ProjectRoleTemplate;
"
`;
